### ISSUE TITLE
Cannot get Azure Keyvault example to work

### ISSUE THREAD

**Issue by iensu** [mozilla/sops#393]
Hi! 

I am trying to follow the Azure Keyvault example (https://github.com/mozilla/sops#encrypting-using-azure-key-vault) but cannot get it to work. I have redone the example several times but always end up with this error:

```
[AZKV]	 ERRO[0030] Encryption failed                             error="azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://sops-XXXXXX.vault.azure.net/keys/sops-key/XXXXXXX/encrypt?api-version=2016-10-01: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Get http://XXX.XXX.XXX.XXX/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fvault.azure.net: dial tcp XXX.XXX.XXX.XXX:80: i/o timeout'" key=sops-key version=a114809b9efe4b7286aa67317a5e5ccd
Could not generate data key: [failed to encrypt new data key with master key "https://sops-XXXXXX.vault.azure.net/keys/sops-key/XXXXXXX": Failed to encrypt data: azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://sops-XXXXXX.vault.azure.net/keys/sops-key/XXXXXXX/encrypt?api-version=2016-10-01: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Get http://XXX.XXX.XXX.XXX/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fvault.azure.net: dial tcp XXX.XXX.XXX.XXX:80: i/o timeout']
```

I am following the instructions step by step, I even tried creating a new resource group and a new service principal, but that didn't work either. I also set the environment variables specified in the instructions to the values provided by the service principal. Any ideas what I'm doing wrong?

**Comment by autrilla** [comment]
You're getting i/o timeouts, are you behind a proxy that may be blocking requests by any chance?

cc @carlpett who wrote the support for this

**Comment by carlpett** [comment]
Hi @iensu,
My guess from your error is that it is for some reason using `http` rather than `https`. Do you have some config which specifies `http://...` directly somewhere, or omits the scheme?

**Comment by iensu** [comment]
Hi @carlpett,

Not that I can see. I'm fairly new to Azure so I probably don't know where to look :)

**Comment by carlpett** [comment]
How far in the instructions do you get before hitting this? To the `sops --encrypt --azure-kv [...]` bit?

I don't really understand why it is trying to refresh a token with a http endpoint... All this should be generated by the Azure SDK. Could you check if you have something in the `.azure` directory in your home directory which has a http://?

**Comment by iensu** [comment]
Hi! Thank you for your swift reply! 

I've looked through all the files in the directory but all URLs are https there.

**Comment by carlpett** [comment]
Hm, strange. I'm following the Azure SDK now and trying to see where things could go wrong... Which `AZURE_*` environment variables do you have set? Are you by any chance running this from within an Azure VM?

**Comment by carlpett** [comment]
I just re-followed the instructions on a freshly downloaded 3.2.0, just to make sure nothing had broken since the older version where it was developed. But it still works for me, so there must be something environment-specific which causes this to happen.
I could build a new version with extra logging if that would be useful? So we can perhaps see where it is going wrong.

**Comment by iensu** [comment]
Yes, that would be great! 

I suspect that there might be some company specific settings interfering here. We've only been tasked with setting up an AKS cluster on a provisioned subnet, so I don't know a lot about the Azure configuration as a whole. 

I'd like to help out though, would be interesting to know what is going wrong. :)

**Comment by carlpett** [comment]
Just realized that I never asked what platform you are on... So here's a zip archive with all of them:
[sops-with-azure-debug.zip](https://github.com/mozilla/sops/files/2573476/sops-with-azure-debug.zip) 
:) 
Right now it just dumps out the config the Azure SDK determines it should use for authentication to stdout. With that info it should be easier to understand what goes wrong.

**Comment by cihangirbesiktas** [comment]
I had the same issue, and also tried with the binary @carlpett provided, and got the following error.
`$ ~/Downloads/sops-with-azure-debug/sops.mac --encrypt --azure-kv https://sops-0bacbd9e7b014cdb.vault.azure.net/keys/sops-key/e1176c2d78b14798b1589ded12bddc13 mysecret.yaml
DEBUG - OauthConfig:  {{      false  } {      false  } {http   169.254.169.254 /metadata/identity/oauth2/token  false api-version=2018-02-01&resource=https%3A%2F%2Fvault.azure.net } {      false  }}
[AZKV]   ERRO[0030] Encryption failed                             error="azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://sops-0bacbd9e7b014cdb.vault.azure.net/keys/sops-key/e1176c2d78b14798b1589ded12bddc13/encrypt?api-version=2016-10-01: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Get http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fvault.azure.net: dial tcp 169.254.169.254:80: i/o timeout'" key=sops-key version=e1176c2d78b14798b1589ded12bddc13
Could not generate data key: [failed to encrypt new data key with master key "https://sops-0bacbd9e7b014cdb.vault.azure.net/keys/sops-key/e1176c2d78b14798b1589ded12bddc13": Failed to encrypt data: azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://sops-0bacbd9e7b014cdb.vault.azure.net/keys/sops-key/e1176c2d78b14798b1589ded12bddc13/encrypt?api-version=2016-10-01: StatusCode=0 -- Original Error: adal: Failed to execute the refresh request. Error = 'Get http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&resource=https%3A%2F%2Fvault.azure.net: dial tcp 169.254.169.254:80: i/o timeout']`

**Comment by cihangirbesiktas** [comment]
It seems the Authorizer uses this code - https://github.com/Azure/azure-sdk-for-go/blob/master/services/keyvault/auth/auth.go and it uses AZURE_KEYVAULT_RESOURCE env variable to set the resource.
Do we need to set AZURE_KEYVAULT_RESOURCE env variable? If so, what should it be? @carlpett

**Comment by carlpett** [comment]
You don't need the `AZURE_KEYVAULT_RESOURCE`, it will default to the correct thing.
For some reason it believes you are both running on the internal Azure network and is checking your identity via the instance metadata endpoints (169.254.169.254), which isn't available on your local workstation (as expected).
The code that does this is [here](https://github.com/Azure/go-autorest/blob/master/autorest/azure/auth/auth.go#L105). The key check is if `AZURE_CLIENT_SECRET` is set, is it in your case?
What is your environment, by the way? (OS/shell/etc)

**Comment by cihangirbesiktas** [comment]
I've set AZURE_CLIENT_SECRET and other azure env vars in the documentation. I am using bash on Mac OS X to run the commands.

**Comment by cihangirbesiktas** [comment]
I also tried on a centos docker instance, and still got the same error.

**Comment by cihangirbesiktas** [comment]
I found the issue, I was setting required env variables as bash variable like below which does not set the env variable.
```
AZURE_CLIENT_SECRET=...
```
The correct way is 
```
export AZURE_CLIENT_SECRET=...
```
So it is better to show the setup of env variables explicitly in the documentation.

**Comment by cihangirbesiktas** [comment]
One other question: How can I setup .sops.yaml file to use azure keyvault?

**Comment by carlpett** [comment]
Ah, yes, that would explain it. Good find! @iensu would this be the cause for you too?
As for the config file, the key is `azure-kv`, and then reference the key the same way as in the command line examples.

**Comment by iensu** [comment]
Sorry! I haven't had the time to try yet, but it definitely looks like it could be. I use fish terminal and I think I only did `set AZURE_CLIENT_SECRET xxx` without passing the export flag. 😟  I will get back to you as soon as I've tried it out!

**Comment by autrilla** [comment]
> So it is better to show the setup of env variables explicitly in the documentation.

I'm not sure we should do this, since we aren't really responsible of how you set your environment variables. What we could definitely do is warn if the variable is set, but empty.

**Comment by iensu** [comment]
It worked! However, now it fails on the decryption stage:

```
DEBUG - TokenURL: https://login.microsoftonline.com/5581c9a8-168b-45f0-abd4-d375da99bf9f/oauth2/token?api-version=1.0
[AZKV]	 INFO[0001] Decryption succeeded                          key=sops-key version=d7804e51b9cd4f19a2ce19391959252b
[SOPS]	 INFO[0001] Data key recovered successfully
[SOPS]	 DEBU[0001] Decrypting tree
Error decrypting tree: Error walking tree: Could not decrypt value: Input string https://login.microsoftonline.com/5581c9a8-168b-45f0-abd4-d375da99bf9f/oauth2/token?api-version=1.0 does not match sops' data format
```

What is happening here?

**Comment by iensu** [comment]
Now it works!

After running through the key creation and policy assignment steps again it now works. My test file was:
```
secretKey: 4649
anotherSecretKey: "hejsan"
```
And that is now working fine, no matter what I do to try and reproduce the error. Unfortunately I had closed my down my terminal window so I can't check if the environment variables where off or what was causing the error. So user error :)

Well, thank you all so much for your help! Closing this issue.
